File: spine/orreth_spine/ground.py
# PROVENANCE: Claude Fable 5.1 (claude-fable-5-1) — rearch intent sp1, the ground at birth · 2026-09-19
# Amended: Claude Fable 5.1 (claude-fable-5-1) — rearch P6 sp4, placement policy v0 · 2026-09-21
# Amended: Claude Fable 5.1 (claude-fable-5-1) — rearch P6.5 sp1, the services ground · 2026-09-22
# Amended: Claude Fable 5.1 (claude-fable-5-1) — rearch P7 sp4, the beat lock (the loops' shadow law) · 2026-09-23
"""Every ground, ensured when a connection is BORN — never inside a serve.

The law (found live at the intent sp1 relight): `outbox.once` keeps a
connection from re-entering DDL, but the FIRST entry happens wherever the
first call lands — and when that is inside a serving transaction (a
resident reading its ask's marker), the DDL's locks live as long as the
serve. A new column on `spine_markers` took an AccessExclusiveLock inside
one serve's savepoint; that serve then waited on the graph, and every
loop and door of the Bridge queued behind a backend "idle in transaction"
for five minutes. So: a rig connection ensures EVERY ground the moment it
is opened, on its own autocommit statements, and no serve ever runs DDL.
Sharpened 2026-09-21 (CI, three runs: two fan-out asks deadlocked a serving
resident at the advisory lock): `outbox.once` keeps a PROCESS memo per
ground (DSN + search_path), so a door's fresh connection to an ensured
ground is born flagged and runs no DDL at all — flagged only once a birth has
FINISHED (the same day, CI: a thread born mid-birth skipped DDL, its seed hit an
undefined table, and the relay died; no resident ever replied).

THE BEAT LOCK (P7 sp4 — the loops' shadow law): two kernels stand on one
ground in SHADOW (the Python Bridge on :4600, the Rust bridge on :4601),
and each runs the standing loops — the scheduler's tick, the intent
rail's turn. A beat reads what is due and acts on it; two beats at once
would file the same occurrence twice, observe the same red twice, plan
the same cause twice. So a beat is CLAIMED before it runs: a session
advisory lock on the ground keyed by the beat's class and this world's
scope (`pg_try_advisory_lock(class, hashtext(scope))`) — held for the
beat, released after; the other kernel's beat finds it held and steps
back, saying so. The lock is Postgres's own: a kernel that dies mid-beat
drops it with its connection. One ground, one beat at a time, per world.
"""
from __future__ import annotations

from contextlib import contextmanager

from . import envelope as ev

BEAT_LOCK = 742200                  # the beat classes count up from here (the DDL guard is 742199)
BEATS = {"scheduler": 1, "intent": 2, "keeper": 3}
HELD = "the beat is held by another kernel on this ground"


def beat_key(name: str) -> int:
    """The advisory lock's first key for a beat class (conformance `beat_lock`)."""
    return BEAT_LOCK + BEATS[name]


def try_beat(conn, name: str) -> bool:
    """Claim this world's beat of a class — True when it is ours to run."""
    cur = conn.cursor()
    cur.execute("SELECT pg_try_advisory_lock(%s::int, hashtext(%s))", (beat_key(name), ev.scope()))
    return bool(cur.fetchone()[0])


def end_beat(conn, name: str) -> None:
    conn.cursor().execute("SELECT pg_advisory_unlock(%s::int, hashtext(%s))",
                          (beat_key(name), ev.scope()))


@contextmanager
def beat(conn, name: str):
    """`with beat(conn, "intent") as ours:` — the beat runs only when ours;
    the lock is released however the beat ends."""
    ours = try_beat(conn, name)
    try:
        yield ours
    finally:
        if ours:
            end_beat(conn, name)

TAGS = ("outbox", "inbox", "projector", "resident", "gateway", "tools", "store",
        "markers", "monitor", "scheduler", "harness", "presence", "digest", "intent",
        "proof", "mitl", "placement", "services")


def ensure_all(conn) -> None:
    """Flag every ground on this connection (each module's `once` tag), so
    the serving transaction never runs DDL and never takes the lock."""
    from . import (digest, gateway, harness, inbox, intent, markers, mitl, monitor, outbox,
                   placement, presence, projector, proof, resident, scheduler, services, store, tools)
    for mod in (outbox, inbox, projector, resident, gateway, tools, store, markers,
                monitor, scheduler, harness, presence, digest, intent, proof, mitl, placement,
                services):
        mod.ensure_schema(conn)
    markers.seed(conn)                      # the kernel's kinds, in this world
    outbox.mark_ground_done(conn, TAGS)     # the birth FINISHED: later connections are born flagged
    proof.seed_masters(conn)                # the SPINE_MASTERS dial, in this world (P6 sp1)


def ensured(conn) -> set[str]:
    """The tags this connection has flagged — the law's evidence."""
    return set(getattr(conn, "_spine_ensured", None) or ())

File: backend/plane/crates/orreth-spine/src/ground.rs
// PROVENANCE: Claude Fable 5.1 (claude-fable-5-1) — rearch P7 sp2, the ground and the rails · 2026-09-22
// Amended: Claude Fable 5.1 (claude-fable-5-1) — rearch P7 sp4, the loops' three tags · 2026-09-23
// Amended: Claude Fable 5.1 (claude-fable-5-1) — rearch P7 sp3, the ask road: the road's six tags at birth · 2026-09-22
//! The ground — mirrors `orreth_spine.ground` and the memo half of
//! `orreth_spine.outbox` (`once` · `ground_key` · `mark_ground_done`).
//!
//! The law (canon 0008's ground law, found live at the intent sp1 relight
//! and sharpened in CI 2026-09-21): DDL runs ONCE per GROUND per PROCESS —
//! never per connection, never inside a serve. A ground is the DSN plus the
//! connection's `search_path` (a test schema and the public one are two
//! grounds). `once(tag)` is true the first time this process sees the tag on
//! this ground; a connection born after a birth FINISHED (`ensure_all` ran to
//! its end and `mark_ground_done` was called) is born flagged and runs no DDL
//! at all — flagged only once the birth has finished, never when a first
//! caller merely claimed the tag (a thread born mid-birth once skipped DDL,
//! hit an undefined table, and the relay died). Every DDL runs inside one
//! transaction holding the advisory lock 742199, the same lock the Python
//! spine takes, so two spines standing on one ground never race a `CREATE`.
//!
//! The tables are THE SAME tables the Python spine uses — same names, same
//! columns, the same `CREATE TABLE IF NOT EXISTS` words — so both spines can
//! stand on one ground in shadow (the sp2 proof).

use crate::rail_error::RailError;
use std::collections::{HashMap, HashSet};
use std::sync::{Mutex, OnceLock};
use tokio_postgres::{Client, NoTls};

/// The DDL race guard — `pg_advisory_xact_lock(742199)`, as Python takes it.
pub const DDL_LOCK: i64 = 742199;

/// The tags this spine ensures at birth (the Python spine ensures more —
/// its `ground.TAGS`; the Rust spine grows its list spoonful by spoonful —
/// sp3 added the ask road's six, sp4 the loops' three).
pub const TAGS: [&str; 16] = [
    "outbox",
    "inbox",
    "heartbeat",
    "resident",
    "markers",
    "proof",
    "intent",
    "presence",
    "digest",
    "store",
    "monitor",
    "scheduler",
    "harness",
    "services",
    "stable",
    "gateway",
];

static GROUNDS_DONE: OnceLock<Mutex<HashMap<String, HashSet<String>>>> = OnceLock::new();

fn memo() -> &'static Mutex<HashMap<String, HashSet<String>>> {
    GROUNDS_DONE.get_or_init(|| Mutex::new(HashMap::new()))
}

/// A birth finished on this ground: connections born later are born flagged
/// for these tags and run no DDL.
pub fn mark_ground_done<'a>(key: &str, tags: impl IntoIterator<Item = &'a str>) {
    let mut m = memo().lock().unwrap_or_else(|p| p.into_inner());
    m.entry(key.to_string())
        .or_default()
        .extend(tags.into_iter().map(str::to_string));
}

/// Has a birth finished on this ground for this tag (the memo's evidence)?
pub fn born_flagged(key: &str, tag: &str) -> bool {
    memo()
        .lock()
        .unwrap_or_else(|p| p.into_inner())
        .get(key)
        .is_some_and(|tags| tags.contains(tag))
}

/// One connection standing on one ground.
pub struct Ground {
    client: Client,
    dsn: String,
    key: String,
    ensured: HashSet<String>,
}

impl Ground {
    /// Open a connection and learn its ground (DSN + `search_path`).
    pub async fn connect(dsn: &str) -> Result<Ground, RailError> {
        let (client, connection) = tokio_postgres::connect(dsn, NoTls).await?;
        tokio::spawn(async move {
            if let Err(e) = connection.await {
                eprintln!("the ground's connection ended: {e}");
            }
        });
        let path: String = client.query_one("SHOW search_path", &[]).await?.get(0);
        Ok(Ground {
            client,
            dsn: dsn.to_string(),
            key: format!("{dsn}|{path}"),
            ensured: HashSet::new(),
        })
    }

    /// Stand on another schema: a NEW ground (new key, nothing ensured yet).
    pub async fn set_search_path(&mut self, path: &str) -> Result<(), RailError> {
        self.client
            .batch_execute(&format!("SET search_path TO {path}"))
            .await?;
        let path: String = self.client.query_one("SHOW search_path", &[]).await?.get(0);
        self.key = format!("{}|{}", self.dsn, path);
        self.ensured.clear();
        Ok(())
    }

    /// The ground this connection stands on — its DSN and its search_path.
    pub fn key(&self) -> &str {
        &self.key
    }

    pub fn client(&self) -> &Client {
        &self.client
    }

    pub fn client_mut(&mut self) -> &mut Client {
        &mut self.client
    }

    /// The tags this connection has flagged — the law's evidence.
    pub fn ensured(&self) -> &HashSet<String> {
        &self.ensured
    }

    /// True the first time this PROCESS sees `tag` on this ground.
    pub fn once(&mut self, tag: &str) -> bool {
        if !self.ensured.insert(tag.to_string()) {
            return false;
        }
        !born_flagged(&self.key, tag)
    }

    /// Run a module's DDL once per ground per process, inside one transaction
    /// holding the advisory lock. Returns whether the DDL ran on this call.
    pub async fn ensure(&mut self, tag: &str, ddl: &[&str]) -> Result<bool, RailError> {
        if !self.once(tag) {
            return Ok(false);
        }
        let tx = self.client.transaction().await?;
        tx.execute("SELECT pg_advisory_xact_lock($1)", &[&DDL_LOCK])
            .await?;
        for stmt in ddl {
            tx.batch_execute(stmt).await?;
        }
        tx.commit().await?;
        Ok(true)
    }

    /// Every ground this spine knows, ensured at BIRTH — never inside a serve
    /// — then the ground marked done so later connections are born flagged.
    pub async fn ensure_all(&mut self) -> Result<(), RailError> {
        crate::outbox::ensure_schema(self).await?;
        crate::inbox::ensure_schema(self).await?;
        crate::heartbeat::ensure_schema(self).await?;
        crate::schema::ensure_road(self).await?;
        mark_ground_done(&self.key, TAGS);
        Ok(())
    }
}
